### PR TITLE
add past data for choral from cassandra

### PR DESCRIPTION
**Pull request by jboretsky**

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -13,8 +13,10 @@ services:
       REDIS_STORE_URI: "redis://redis:6379"
       REDIS_STORE_SECRET: "DEVELOPMENT_SECRET"
       MONGO_URL: "mongodb://mongodb:27017"
+      CASSANDRA_URL: "cassandra"
```
Comment (andreirtaylor) [inline docker-compose.yml:16]: We will need to update cluster.yml with this.

File: routes/chorals.js [hunk 2]
```diff
@@ -72,26 +74,70 @@ router.get('/new', function(req, res, next) {
...
+          }
+        }
+        ret.tabs = tabs;
+        resolve(ret);
+      });
+    });
+  }).then((ret) => {
+    return new Promise((resolve, reject) => {
+      cass_client.execute("SELECT * FROM choraldatastream.raw_data WHERE device_id = '" + choralId + "' order by device_timestamp DESC limit " + 3600/ret.sampleRate, function( err, result ) {
+        var sorted_results = {};
```
Comment (andreirtaylor) [inline routes/chorals.js]: result could be null. Perform a check here

File: routes/chorals.js [hunk 3]
```diff
@@ -72,26 +74,70 @@ router.get('/new', function(req, res, next) {
...
+      ret.sampleRate = choral.sampleRate;
+      resolve(ret);
     });
   });
 
-  getTabs.then((tabs) => {
+  p.then((ret) => {
+    return new Promise((resolve,reject) => {
+      client.hgetall(choralId, (err, data) => {
+        var tabs = [];
```
Comment (andreirtaylor) [inline routes/chorals.js]: data could be null and there could be an error

File: routes/chorals.js [hunk 4]
```diff
@@ -72,26 +74,70 @@ router.get('/new', function(req, res, next) {
...
+        resolve(ret);
+      });
+    });
+  }).then((ret)=> {
     res.render('choral', {
       googleUser: req.user,
       parentChoralId: req.params.choralId,
-      tabs: tabs
+      tabs: ret.tabs,
+      past_data: ret.past_data,
```
Comment (andreirtaylor) [inline routes/chorals.js]: ret could be null here.

File: routes/chorals.js (hunk 3 above)
Comment (jboretsky) [reply routes/chorals.js]: I guess we could see an error if sample rate is very long and we try accessing this before there's any data.

File: routes/chorals.js [hunk 5]
```diff
@@ -97,26 +99,83 @@ router.get('/new', function(req, res, next) {
...
+  // 2. default currently is 1 hour of data, so get (3600/sampleRate) data points
+  // 3. put them into object where {tabName: [data_points]} so FE doesn't need to do any work
+  let p = new Promise((resolve, reject) => {
+    ret = {};
+    Choral.findOne({ choralId: choralId }, (err, choral) => {
+      if (err) {
+        logHelper.createLog("error", 'Choral does not exist: ' + err, ["chorals", "get"]);
+        console.log(err);
+        res.flash('error', 'Choral does not exist');
+        return res.send('404'); // notify client of failure
```
Comment (cpjk) [inline routes/chorals.js]: does this work?

Comment (cpjk) [reply routes/chorals.js]: i thought you would have to reject(error reason) and handle that outside the promise declaraation

Comment (jboretsky) [reply routes/chorals.js]: hmmm I think you're right.

Comment (jboretsky) [reply routes/chorals.js]: promise.catch()

Comment (jboretsky) [reply routes/chorals.js]: I just tried it actually, and it does work. I think res.send just overrides the promise chain and redirects you.

Comment (jboretsky) [reply routes/chorals.js]: but I need to check if(err || !choral) so let me update that

Comment (cpjk) [reply routes/chorals.js]: coo

File: routes/chorals.js [hunk 6]
```diff
@@ -97,26 +99,83 @@ router.get('/new', function(req, res, next) {
...
+            }
+          }
+        }
+        ret.tabs = tabs;
+        resolve(ret);
+      });
+    });
+  }).then((ret) => {
+    return new Promise((resolve, reject) => {
+      cass_client.execute("SELECT * FROM choraldatastream.raw_data WHERE device_id = '" + choralId + "' order by device_timestamp DESC limit " + 3600/ret.choralInfo.sampleRate, function( err, result ) {
```
Comment (cpjk) [inline routes/chorals.js]: lol injection babay

Comment (cpjk) [reply routes/chorals.js]: eff it

Comment (jboretsky) [reply routes/chorals.js]: hhahhaha wow that's awesome

Comment (jboretsky) [reply routes/chorals.js]: hmm yeah oh well for now

Comment (cpjk) [inline routes/chorals.js]: Probably want an error check in here

**Comment by cpjk** [comment]
yolo

**Pull request merged**